### DISCUSSION TITLE
Storage RLS on specific folder with a table join

### DISCUSSION THREAD

**Discussion by unlocomqx** [supabase/supabase#28160] [Questions]
Hi,
I've been trying to create a RLS policy involving a table join and a folder name but without luck

Here's an example (the bucket name is `forms`)
File path:
`1/tables/2/sample.csv` 

In my db, I have two tables: `forms` and `tables`

auth.uid() -> `forms.id_user` -> `tables.id_form`

I wanted to check inserts against the id_user and the tables folder id (in this case, it's `2`)

Here's what I tried

```sql
(bucket_id = 'forms'::text) AND ( SELECT auth.uid() AS uid) IN (
   SELECT forms.id_user FROM tables
   LEFT JOIN forms
   ON tables.id_form = forms.id
   WHERE tables.id = (storage.foldername(name))[2]::integer
)
```
But I had this error

```txt
Error updating policy: failed to update pg.policies with the given ID: column reference "name" is ambiguous
```
How to include the `storage.foldername(name)` part in the query? It's probably not possible with a join?

I also tried without a join

```sql
(bucket_id = 'forms'::text) AND ( SELECT auth.uid() AS uid) IN (
   SELECT id_user FROM forms
   WHERE id = (SELECT id_form FROM tables WHERE id = (storage.foldername(name))[2]::integer)
)
```

which saved with success but gave me and error in my app

`Error: new row violates row-level security policy`

I suspect that my folder index is wrong? The documentation is not very clear to me

What is the folder index of `2` in this case: `1/tables/2/sample.csv` (I tried 2 and 3 but it didn't work anyway)

How to debug this? Is there a log I can check that shows the raw query and the query result?
What am I doing wrong?

Thank you very much

Edit: I noticed that supabase is transforming `storage.foldername(name)` into `storage.foldername(tables.name)`. How to avoid this?

Edit: I also tried this but it also gave me a row security error

```sql
(bucket_id = 'forms'::text) AND (storage.foldername(name))[2]::integer IN ( SELECT tables.id
   FROM (tables
     LEFT JOIN forms ON ((tables.id_form = forms.id)))
  WHERE forms.id_user = auth.uid())
```

**Comment by GaryAustin1** [comment]
Postgres numbers arrays from 1 so you want 3.
Your name error is likely because you have a name column in forms and it does not know which you are referring to.   So something like foldername(storage.objects.name) is likely needed.

**Reply by unlocomqx** [reply]
Thank you your answer. I used `objects.name` and the syntax went trough fine. But the rule didn't allow me to insert a file. I tried everything.

I ended up using a definer function, something like

```sql
create function private.table_belongs(table_id text)
returns boolean
language plpgsql
security definer -- will run as the creator
as $$
begin
  return exists (
    select 1 from tables
    JOIN forms ON forms.id = tables.id_form
    where (select auth.uid()) = forms.id_user AND tables.id = table_id::int4
  );
end;
$$;
```

Cheers :wave:

**Reply by GaryAustin1** [reply]
If security definer worked then most likely you were not meeting RLS on the other tables.

**Reply by unlocomqx** [reply]
I still didn't enable rls on the public schema

![CleanShot 2024-07-24 at 17 26 08@2x](https://github.com/user-attachments/assets/056576c9-eb9c-423a-8ee8-4a140fa4c178)

probably a rls issue with the storage table itself

![CleanShot 2024-07-24 at 17 28 38@2x](https://github.com/user-attachments/assets/7e7ad6cf-54f5-47e1-8ce3-172a08cf14b7)

It will take me some time to wrap my head around rls (0.1x engineer here :wave:)
